### PR TITLE
Tool to sniff data contents before encrypted with OpenSSL

### PR DESCRIPTION
**Pull request by adrianlzt**
Add tool as talked in iovisor-dev 'BCC: bpf_probe_read read function arguments'

### CODE REVIEW

**Comment by adrianlzt** [comment]
@markdrayton I take your version and clean it a little

**Comment by markdrayton** [comment]
Excellent, thanks! I have a few review remarks — will add them inline :)

Also, can you make a man page? It should be easy to copy and modify an example from `man/man8`.

File: tools/sslsniff.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        perf_SSL_read.perf_submit(ctx, &__data, sizeof(__data));
+        return 0;
+}
+"""
+
+b = BPF(text=prog)
+libpath = "/usr/lib/libssl.so.1.0.0"
+
+# Join to ssl_write
+b.attach_uprobe(name=libpath, sym="SSL_write", fn_name="probe_SSL_write")
```
Comment (markdrayton) [inline tools/sslsniff.py]: You can just use `name="ssl"` here. It'll be resolved into the correct path to libssl (which may differ from yours) automagically.

File: tools/sslsniff.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+{
+        u64 timestamp_ns;
+        u32 pid;
+        char comm[TASK_COMM_LEN];
+        char v0[476];
+};
+
+BPF_PERF_OUTPUT(perf_SSL_write);
+
+int probe_SSL_write(struct pt_regs *ctx)
```
Comment (markdrayton) [inline tools/sslsniff.py]: In my quick test it appears that listing args on the function definition works, e.g.:

    int probe_SSL_write(struct pt_regs *ctx, void *ssl, void *buf, int num) {

(same for the other probes)

This is (IMO) clearer than using the PT_REGS_* macros to access the function args.

File: tools/sslsniff.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        __data.timestamp_ns = bpf_ktime_get_ns();
+        __data.pid = bpf_get_current_pid_tgid();
+
+        bpf_get_current_comm(&__data.comm, sizeof(__data.comm));
+
+        int ret;
+
+        if ( (void *) PT_REGS_PARM2(ctx) != 0) {
+                bpf_probe_read(&__data.v0, sizeof(__data.v0), (void *) ctx->si);
+                ret = bpf_probe_read(&__data.v0, sizeof(__data.v0), (unsigned char*) PT_REGS_PARM2(ctx));
```
Comment (markdrayton) [inline tools/sslsniff.py]: Do we need both `bpf_probe_reads`? Also, `ret` doesn't seem to be used.

File: tools/sslsniff.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        struct probe_SSL_data_t __data = {0};
+        __data.timestamp_ns = bpf_ktime_get_ns();
+        __data.pid = pid;
+
+        bpf_get_current_comm(&__data.comm, sizeof(__data.comm));
+
+        int ret;
+        if (bufp != 0) {
+                ret = bpf_probe_read(&__data.v0, sizeof(__data.v0), (char *)*bufp);
+        }
```
Comment (markdrayton) [inline tools/sslsniff.py]: same for `ret`.

**Comment by markdrayton** [comment]
It'd also be nice if there was an indicator when data is truncated, just so people aren't surprised that only the first N bytes are captured.

Also, @4ast, are there any clever ways to work around the stack size limitation when sending data back through perf?

**Comment by markdrayton** [comment]
One last request: can you PEP8 format the code and add `from __future__ import print_function` for consistency with the other tools?

File: tools/sslsniff.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        u32 pid = bpf_get_current_pid_tgid();
+        u64 *bufp = bufs.lookup(&pid);
+        if (bufp == 0) {
+                return 0;
+        }
+
+        struct probe_SSL_data_t __data = {0};
+        __data.timestamp_ns = bpf_ktime_get_ns();
+        __data.pid = pid;
+        __data.len = num;
```
Comment (markdrayton) [inline tools/sslsniff.py]: `num` is the number of bytes the user asked for, which isn't always the number read. The return value of `SSL_read` (e.g. `PT_REGS_RC(ctx)`) will give you the number read.

File: tools/sslsniff.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        start = event.timestamp_ns
+    time_s = (float(event.timestamp_ns - start)) / 1000000000
+
+    s_mark = "-" * 5 + " DATA " + "-" * 5
+
+    e_mark = "-" * 5 + " END DATA " + "-" * 5
+
+    # If data length is exactly MAX_BUF_SIZE, it will be reported
+    # as truncated when it is not
+    if len(event.v0) == MAX_BUF_SIZE:
```
Comment (markdrayton) [inline tools/sslsniff.py]: Use `if event.len > MAX_BUF_SIZE` (after changing `len` as above) to accurately tell if we truncated.

File: tools/sslsniff.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+int probe_SSL_write(struct pt_regs *ctx, void *ssl, void *buf, int num) {
+        struct probe_SSL_data_t __data = {0};
+        __data.timestamp_ns = bpf_ktime_get_ns();
+        __data.pid = bpf_get_current_pid_tgid();
+        __data.len = num;
+
+        bpf_get_current_comm(&__data.comm, sizeof(__data.comm));
+
+        if ( (void *) buf != 0) {
```
Comment (markdrayton) [inline tools/sslsniff.py]: I think you can drop the `void *` casts here and below.

File: man/man8/sslsniff.8 [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+Print all calls to SSL_write and SSL_read system-wide:
+#
+.B sslsniff
+.SH FIELDS
+.TP
+FUNC
+Which function is being called (SSL_write or SSL_read)
+.TP
+TIME
+Time of the command (HH:MM:SS).
```
Comment (markdrayton) [inline man/man8/sslsniff.8]: nit: the timestamp is in seconds, not HH:MM:SS (one day we should make timestamps in all the tools consistent!)

**Comment by markdrayton** [comment]
Looking good! Once the other things are finished can you squash this down to one commit and force push the branch?

**Comment by brendangregg** [comment]
Nice! Please add a link in the tools list in README.md so people can quickly find the tool and example.

**Comment by adrianlzt** [comment]
Thanks for the help improving this!
Will be great some _trick_ to be able to capture more data.

**Comment by brendangregg** [comment]
ok to test

**Comment by brendangregg** [comment]
LGTM, merging, thanks!

**Pull request merged**